### ISSUE TITLE
About Wav2Vec(and VQ-wav2vec) objective function

### ISSUE THREAD

**Issue by Moonjong-Shin** [facebookresearch/fairseq#4823]
#### What is your question?

the loss function looks like slightly different.
there's no sigmoid function with True future and negative sample and linear projection layer

Is it happen cause of my misunderstanding of einsum or not implemention this detail?

It would be perfect, if somone answer about this code section. 

Thanks
![image](https://user-images.githubusercontent.com/104052651/197656639-c5f0251b-e8af-41c0-98f5-7ab7dd3f2e2d.png)

#### Code
[here](https://github.com/facebookresearch/fairseq/blob/0272196aa803ecc94a8bffa6132a8c64fd7de286/fairseq/models/wav2vec/wav2vec.py#L610)
```python
if self.infonce:
                predictions[start:end] = torch.einsum(
                    "bct,nbct->tbn", x[..., :-offset, i], targets[..., offset:]
                ).flatten()
```

#### What's your environment?

 - fairseq Version (e.g., 1.0 or main): 0.12.2
 - PyTorch Version (e.g., 1.0) : 1.12.1
 - OS (e.g., Linux): Window 11
 - How you installed fairseq (`pip`, source): pip install fairseq
 - Python version: 3.8
 - CUDA/cuDNN version: 11.6
 - GPU models and configuration: NVIDIA Grforce 3090 Ti

**Comment by Moonjong-Shin** [comment]
sorry about my misunderstanding.

the output is defined as `cpc_logit` and if the negative sample set is defined as 10,
the `cpc_logti`'s shape is (n, 11)  11 is from one positive sample(true future samples) and 10 negative samples

so the loss function have to be calculated outof the model 

maybe the objective function is like

```python 

sigmoid = nn.Sigmoid()

true = result['cpc_logit'][..., 0]
negatives = result['cpc_logit'][..., 1:]

loss  = -(torch.log(sigmoid(true) + torch.log(sigmoid(-negatives))))

loss.backward()
```

would be so nice if someone check about this objectivbe function code
